src/vhip/mlmodel/gene_features.py

```diff
@@ -160,13 +160,58 @@ def __init__(self, gene_file: str) -> None:
             raise Exception(
                 "Genes file is not provided or empty. Please provide a valid gene file."
             )
-
+        self.id = gene_file
+        self.genes: List[Gene] = []
+        self.skipped_genes: List[str] = []
         readout = read_annotated_genes(gene_file)
-        self.genes: List[Gene] = [
-            Gene(
-                gene_seq=str(readout[0][out]),
-                gene_id=str(readout[1][out]),
-                gene_product=str(readout[2][out]),
+
+        for out in range(len(readout[0])):
+            try:
+                self.genes.append(
+                    Gene(
+                        gene_seq=str(readout[0][out]),
+                        gene_id=str(readout[1][out]),
+                        gene_product=str(readout[2][out]),
+                    )
+                )
+            except Exception:
+                self.skipped_genes.append(str(readout[1][out]))
+        print(f"{len(self.skipped_genes)} genes skipped")
+
+    def codon_counts(
+        self, threshold_imprecise: int, threshold_skipped_genes: int
+    ) -> None:
+        """Aggregate the counts for each unique codon and imprecise codons across an entire GeneSet.
+
+        Args:
+            threshold_imprecise (int): Number of imprecise (non-ATGC) codons tolerated in a single gene.
+            threshold_skipped_genes (int): Tolerated number of genes in GeneSet that have more than threshold_imprecise codons.
+        Populates the following class attributes:
+            self.codon_dict (str: int): Counts of each unique codon across all genes in the GeneSet.
+            self.imprecise_codons (int): Total number of imprecise codons found in the GeneSet.
+            self.skipped_imprecise_genes (List(str)): IDs of genes in the GeneSet that have more than threshold_imprecise codons.
+        """
+        self.codon_dict = dict.fromkeys(CODON_LIST, 0)
+        self.imprecise_codons: int = 0
+        self.skipped_imprecise_genes: List[str] = []
+
+        counter = 0
+        for gene in self.genes:
+            counter += 1
+            print(f"Analyzing gene 1 of {len(self.genes)}")
+            gene.calculate_codon_counts()
+            self.imprecise_codons += gene.number_imprecise_codons
+            if gene.number_imprecise_codons <= threshold_imprecise:
+                for key, val in gene.codon_dict.items():
+                    self.codon_dict[key] += val
+            else:
+                self.skipped_imprecise_genes.append(gene.gene_id)
+
+        if len(self.skipped_imprecise_genes) > threshold_skipped_genes:
+            raise Exception(
+                f"Too many skipped genes. {len(self.skipped_imprecise_genes)} genes have > {threshold_imprecise} imprecise codons."
+            )
+        elif len(self.skipped_imprecise_genes) > 0:
+            print(
+                f"Skipped {len(self.skipped_imprecise_genes)} genes with too many imprecise codons"
             )
```

Comment:
Is there a way to retrieve those skipped genes? have an attribute or method to get that information back would be super useful I think

Reply:
So, both the skipped_genes (intented to be genes skipped for length not divisible by codon length) and the skipped_imprecise_genes (genes with too many imprecise codons) attributes store the gene ids. Were you wanting to retrieve other information too? Like a method that returns a list of gene products and/or gene sequences using the list in the skipped genes attribute(s)?

Reply:
gene ids is good enough! just a way for a user to lookup those missed genes

Reply:
Got it. So--leave as is? The user can access those the attribute of any GeneSet object?